### ISSUE TITLE
Parallelization of parameter evaluation

### ISSUE THREAD

**Issue by gvnjones** [Qiskit/qiskit-terra#5747]
Performing a standard VQE-qUCCSD calculation for a chemical system requiring ~20 qubits takes many dozens of hours using Qiskit for several computational platforms and simulators. By comparison, performing RHF, MP2, CCSD,  DMRG or FCI for a system of 20 spin-orbitals on a classical computer requires ~1 minute. This severely slows down research on quantum applications in quantum chemistry and materials science.

The issue appears to be related to the fact that Qiskit parameter evaluations are performed in serial, rather than in parallel. Qiskit contains optimizers that are taken straight from SciPy and do not support parallel gradient evaluation. These optimizers need to be rewritten to enable parallelization of the evaluation of gradient components. 

Such an issue can be evaluated with the code shown below involving the simulation of a molecule requiring 22 qubits with VQE-qUCCS, which has greater than 1400 parameters and takes ~300s (5 mins) per evaluation on a simulator:

```
from qiskit                                        import *
from qiskit.chemistry.drivers                      import PySCFDriver, UnitsType,HFMethodType
from qiskit.chemistry.core                         import Hamiltonian,TransformationType,QubitMappingType
from qiskit.chemistry.components.initial_states    import HartreeFock
from qiskit.chemistry.components.variational_forms import UCCSD
from qiskit.aqua.components.optimizers             import L_BFGS_B,COBYLA
from qiskit.aqua.algorithms                        import VQE
from qiskit.aqua                                   import QuantumInstance,aqua_globals

import logging
from   qiskit.chemistry import set_qiskit_chemistry_logging
set_qiskit_chemistry_logging(logging.DEBUG)

driver    = PySCFDriver(atom='''S 0.0000 0.0000 0.1030; Li 0.0000 0.9616 -0.8239; Li 0.0000 -0.9616 -0.8239''',
                        unit=UnitsType.ANGSTROM,charge=0,spin=0,basis='sto-6g',hf_method=HFMethodType.RHF)
molecule  = driver.run()

orb_red   = [0,1,2,3,4,5,6] # simple frozen core
core      = Hamiltonian(transformation=TransformationType.FULL,qubit_mapping=QubitMappingType.PARITY,
                        two_qubit_reduction=True,freeze_core=False,orbital_reduction=orb_red)
H_op,A_op = core.run(molecule)

z2syms,sqlist           = None,None
init_state = HartreeFock(num_orbitals=core._molecule_info['num_orbitals'],qubit_mapping=core._qubit_mapping,
                         two_qubit_reduction=core._two_qubit_reduction,num_qubits=H_op.num_qubits,num_particles=core._molecule_info['num_particles'],sq_list=sqlist)
circuit    = init_state.construct_circuit()

var_form  = UCCSD(num_orbitals=core._molecule_info['num_orbitals'],num_particles=core._molecule_info['num_particles'],active_occupied=None,active_unoccupied=None,
                  initial_state=init_state,depth=1,num_qubits=H_op.num_qubits,qubit_mapping=core._qubit_mapping,two_qubit_reduction=core._two_qubit_reduction,num_time_slices=1,z2_symmetries=z2syms)

optimizer = L_BFGS_B(maxiter=1000)
algo      = VQE(H_op,var_form,optimizer,aux_operators=A_op)

backend          = Aer.get_backend('qasm_simulator')
quantum_instance = QuantumInstance(backend=backend,shots=1)
algo_result      = algo.run(quantum_instance)
```

{'qiskit-terra': '0.14.2', 'qiskit-aer': '0.5.2', 'qiskit-ignis': '0.3.3', 'qiskit-ibmq-provider': '0.7.2', 'qiskit-aqua': '0.7.3', 'qiskit': '0.19.6'}

**Comment by nonhermitian** [comment]
Is there a profiling anywhere to show that it is the optimizer that it is the bottleneck here?  1400 doubles should not be too bad for the Fortran based solver used in the example.

**Comment by woodsp-ibm** [comment]
During gradient computation a number of points in the locale are computed (one in each dimension). Since each point needs an energy evaluation, and many of these are needed to do a gradient, the idea is to be able to do these in parallel when doing simulation. So rather than computing say 100 evals in sequence for 100 points to do a single gradient, do these in parallel if possible when using the classical simulation. So the bottleneck is more the simulation needed than the optimizer itself.

**Comment by woodsp-ibm** [comment]
This is rather old now and things have changed considerably over the years since, with primitives being the main execution point now. VQE still supports max_evals_grouped, which though must be now set directly on the optimizer being used. This will result in a call to the Estimator with a list of points - at this stage what and Estimator does with that in terms of execution is down to the provider, that supplied the primitive, and its implementation. 

I know I indicated this in the past too https://medium.com/qiskit/improve-the-performance-of-your-multi-circuit-simulations-using-qiskit-aer-with-dask-clusters-b4e47fad6cf0 which could be of interest.

Anyway since this is so outdated, and with algorithms also being imminently moved out of Qiskit, I am going to close this. If you feel its something that is still needed maybe a new issue that is more contemporary - maybe more geared towards Aer and the Aer primitives if that's the environment/parallelization desired. With the Estimator its also arguably easier nowadays to roll your own VQE so maybe there is some way to parallelize things more as an end user by having something more custom.
